### ISSUE TITLE
Search problem from Turkish to English

### ISSUE THREAD

**Issue by Guybrush88** [Tatoeba/tatoeba2#1296]
As reported by DostKaplan on the wall:

> [Bug] Search from Turkish to English...
> 
> "iyi olur" returns 948 results.
> "iyi Olur" returns 948 results. (So clearly upper or lower case doesn't matter.)
> 
> But...
> 
> "İyi olur" (note the capital İ Turkish character) returns 0 results:
> 
> No results found for: "İyi olur"

https://tatoeba.org/ita/wall/show_message/27068#message_27068

**Comment by alanfgh** [comment]
This can be fixed via the index. In that Wall thread, I asked Turkish speakers to specify all sets of characters that should be treated identically for the purpose of search.

**Comment by alanfgh** [comment]
See new Wall thread: 

https://tatoeba.org/eng/wall/show_message/27292#message_27292

**Comment by alanfgh** [comment]
I tried fixing this, but it's not working correctly. Not sure whether this is due to my testing procedure or my code change.

Here's the code I added at line 255 of app/vendors/shells/sphinx_conf.php:

```
        /* Turkish
         *   Vowels:
         *     Fold dotless capital I into lowercase dotless i ('I->U+131')
         *     Fold dotted capital I into lowercase dotted i ('U+130->i')
         *     Fold {A,a,U,u} + ^ (U+C2, U+E2, U+DB, U+FB) into letters without ^
         *     Fold {I,i} + ^ (U+CE, U+EE) into lowercase dotless i without ^
         *     Rely on Latin-1 Supplement rules for O, o, U, u with/without umlaut
         *   Consonants:
         *     Fold G + breve (U+11E) into g + breve (U+11F)
         *     Fold S + cedilla (U+15E) into s + cedilla (U+15F)
         *     Rely on Latin-1 Supplement rules for C, c with/without cedilla
         *   Combining marks:
         *     Ignore combining circumflex (U+302)
         */
        $this->indexExtraOptions['tur'] = 
            "
        charset_table = ".implode(', ', array_merge(
            array(
                'A->a', 'U+00C2->a', 'U+00E2->a', 'a',
                'B..H->b..h', 'b..h',
                'I->U+131', 'U+00CE->U+131', 'U+00EE->U+131', 'U+131',
                'U+130->i', 'i',
                'J..T->j..t', 'j..t',
                'U->u', 'U+00DB->u', 'U+00FB->u', 'u',
                'V..Z->v..z', 'v..z',
                'U+11E->U+11F', 'U+11F',
                'U+15E->U+15F', 'U+15F',
            ),
            array_filter(
                $this->charsetTable,
                function($v) { 
                    return $v != 'U+302'
                    && $v != 'A..Z->a..z' && $v != 'a..z'
                    && $v != 'U+00C2->U+00E2' && $v != 'U+00E2'
                    && $v != 'U+00CE->U+00EE' && $v != 'U+00EE'
                    && $v != 'U+00DB->U+00FB' && $v != 'U+00FB'
                ; }
            )
        ))."\n";

```
Here's what I did:

```
.\cake\console\cake -app app sphinx_conf > sphinx.conf
move sphinx.conf c:\code\Tatoeba\sphinxdata\sphinx.conf
indexer --all -c C:\code\tatoeba\sphinxdata\sphinx.conf
[I stopped and restarted MySQL and Apache from the XAMPP Control Panel.]
searchd -c c:\code\Tatoeba\sphinxdata\sphinx.conf
```
Results for searches with "Turkish" as the source language and "Any language" as the target:

"^İyi" finds the same results as "^Iyi". It shouldn't.
"hâlâ" misses sentences with "hala". It shouldn't.
"hala" misses sentences with "hâlâ". It shouldn't.

The handling of vowels with umlauts and consonants with cedillas and breves works correctly with my changes, but it was working even before my changes.

Any idea what I'm doing wrong?

**Comment by trang** [comment]
Seems like it is a problem in the code but I haven't worked enough with Sphinx to tell you where the problem is. I would need to invest maybe a whole weekend on this but this is not high enough priority for me to afford to invest so much time.

One thing though, for this issue it would be really good to have a set of sentences with predefined test cases.

For instance I started with testing the case DostKaplan mentioned in his original post. I started from an empty database and added the following sentences:

* Onu yemesem iyi olur.  
* Şimdi başlasan iyi olur.
* Uyanık olsan iyi olur.

I searched `"iyi olur"`, all these sentences were returned.
I searched `"İyi olur"` (capitalized the first `i` into `İ`), no results were returned. Expected: all the sentences above should also be returned.

These test cases should be confirmed by Turkish speakers so that we make sure we're not misunderstanding something.
They will also be very useful for any developer who doesn't know Turkish and would like to help on this issue. It personally took me quite a while to figure out what data I should gather to test your code.

-----

Side note: you don't need to restart MySQL and Apache after indexing.

**Comment by alanfgh** [comment]
> It personally took me quite a while to figure out what data I should gather to test your code.

Sorry about that. I had some private correspondence with DostKaplan and was using test cases that were based on both what he suggested there and what he suggested in his original message. But since the cases I mentioned were enough to find problems in the search in my database (your local database, which you mention in section 5.2 of this page: https://github.com/Tatoeba/tatoeba2/wiki/How-to-install-Tatoeba ), I thought I didn't need to go into detail beyond what I wrote above.

In the meantime, gillux suggested some things for me to try and asked for a copy of my sphinx.conf, so I sent it to him (along with the sphinx_conf.php that I modified to generate it).

**Comment by jiru** [comment]
I think there is a problem with these lines:
```php
                    return $v != 'U+302'
                    && $v != 'A..Z->a..z' && $v != 'a..z'
                    && $v != 'U+00C2->U+00E2' && $v != 'U+00E2'
                    && $v != 'U+00CE->U+00EE' && $v != 'U+00EE'
                    && $v != 'U+00DB->U+00FB' && $v != 'U+00FB'
```
Most do not serve the purpose you want them to. Everything you put in there need to exist in `$this->charsetTable`. But there is no 'U+302' or 'U+00C2->U+00E2' in `$this->charsetTable`. Only 'A..Z->a..z' and 'a..z' are valid.

What we’re doing here is taking `$this->charsetTable`, removing some of its elements with a filter function (for instance 'A..Z->a..z' and 'a..z') and then adding some more elements (like `'A->a', 'U+00C2->a', 'U+00E2->a', 'a', 'B..H->b..h', 'b..h' etc.`) for a particular language.

So if you want, say, to fold some characters into U+131, then you first need to remove the existing 'U+100..U+177/2' and replace it with 'U+100..U+130/2' and 'U+132..U+177/2'. The same goes for the other characters you want to special-case. Otherwise, you keep the existing mappings along with the ones you’re adding.

I admit the way we manipulate `$this->charsetTable` here is rather clumsy and error-prone, and it could be improved. But, as always, it’s because it all started as a hack.

**Comment by alanfgh** [comment]
Aha! So the ranges in the filter need to match exactly how they're defined in ```charset_Table```. It's not good enough for the range in ```charset_Table``` to contain the range in the other. I've changed this locally and now my search is working correctly. Thanks a lot!

**Comment by alanfgh** [comment]
It's working now. I'm going to send the updated file (which also contains a fix for #1088) to Trang.

**Comment by alanfgh** [comment]
Capitalized and lowercase dotless I should be treated as identical when searching in Turkish.

To test, I added the following artificial "sentences" to my local corpus.

Group 1:

cap-dotless-i IIII
small-dotless-ı ıııı

A search in Turkish for "IIII" or "ıııı" should find both of these "sentences".

:::

Capitalized and lowercase dotted i should be treated as identical when searching in Turkish. As DostKaplan put it:
:::
"iyi olur", "iyi Olur", "İyi olur" should result in the same search and return the same result set.
:::

To test, I added the following artificial "sentences" to my local corpus.

Group 2:

cap-dotted-i İİİİ
small-dotted-i iiii

A search in Turkish for "İİİİ" or "iiii" should find both of these "sentences".

:::

DostKaplan wrote this:
:::
While Â/â, Î/î, and Û/û occur rather infrequently in Turkish, they are sometimes used. For example the word "hâlâ" (which means "still" as in "He's still talking") is often spelled without the accents, both of which return non-zero results in Tatoeba. Neither are wrong because Turks spell it both ways, with the accented version appearing mostly in literary work. But "hala" (without accents) also means "paternal aunt." In my opinion, since Â/â, Î/î, and Û/û occur so infrequently, they should be relaxed in the search, so that searching for "hala" also returns sentences containing "hâlâ", but I am not sure what algorithm you would use to achieve this optimally and effectively.
:::

Accordingly, I modified the index to treat A/a with and without circumflex as identical in a search within Turkish. I did similarly for I/i, and for U/u.

To test, I added the following artificial "sentences" to my local corpus.

Group 3:
small-a-circumflex ââââ
cap-a-circumflex ÂÂÂÂ
small-a aaaa
cap-a AAAA

A search for "ââââ", "ÂÂÂÂ", "aaaa", or "AAAA" should find all of these "sentences".

Group 4:

small-i-circumflex îîîî
cap-i-circumflex ÎÎÎÎ
small-i iiii
cap-i IIII

A search for "îîîî", "ÎÎÎÎ", "iiii", or "IIII" should find all of them.

Group 5:

small-u-circumflex ûûûû
cap-u-circumflex ÛÛÛÛ
small-u uuuu
cap-u UUUU

A search for "ûûûû", "ÛÛÛÛ", "uuuu", or "UUUU" should find all of them.

In addition, I made sure that these words (some with artificial capitalizations) returned the same result sets:

iyi
İyi

ıyi
Iyi

hâlâ
hÂlâ
hala

ağaç
aĞaç
aĞaÇ

bağırıyor
bağIrıyor
baĞırıyor

çok
Çok

sürüyor
sÜrüyor

öptü
Öptü

şükür
Şükür
